File: network/Blockchain.cpp
#include "Blockchain.h"
#include <algorithm>
#include <cmath>
#include <iostream>
#include <memory>

Blockchain::Blockchain(const std::string& chain_name, IMempool* mempool)
        : chain(chain_name), difficulty(1), mempool(mempool), adjustment_interval(2016) {
    if (!mempool) {
        std::cerr << "Aviso: Mempool é nullptr. Blockchain não terá acesso à mempool." << std::endl;
    }
}

double Blockchain::GetDifficulty() const {
    return this->difficulty;
}

void Blockchain::AddBlock(const IBlock& block) {
    if (block.IsValid()) {
        blocks.Add(block.Clone());
        AdjustDifficulty();
    }
}

bool Blockchain::CheckChainIntegrity() const {
    bool status = true;

    for (int i = 0; i < blocks.Size(); ++i) {
        auto currentBlock = blocks.Get(i);
        if (!currentBlock || !currentBlock.get()->IsValid()) {
            status = false;
            break;
        }
    }
    return status;
}

std::shared_ptr<IBlock> Blockchain::GetLastChainBlock() const {
    return (blocks.Size() > 0) ? blocks.Get(blocks.Size() - 1) : nullptr;
}


void Blockchain::DisplayChain() const {
    std::cout << "===================[ BLOCKCHAIN INFO ]=====================" << std::endl;
    std::cout << "Total Blocks in the blockchain: " << this->blocks.Size() << std::endl;
    std::cout << "Current Difficulty:  " << this->difficulty << std::endl;
    std::cout << "Total Transactions pending in the mempool: " << this->mempool->GetPendingTransactions().size() << std::endl;
    std::cout << "===========================================================" << std::endl;
    /*
    for (int i = 0; i < blocks.Size(); ++i) {
        auto block = blocks.Get(i);
        if (block) {
            auto blockData = block.get();
            std::cout << "Block #" << i << " => Hash: " << blockData->GetHash()
                      << ", Previous Hash: " << blockData->GetPreviousBlockHash()  << " Timestamp: " << blockData->GetTimestamp() << std::endl;
        }
    }
     */
}

void Blockchain::AdjustDifficulty() {
    int blockCount = this->blocks.Size();

    if (blockCount % adjustment_interval == 0 && blockCount > 0) {
        std::vector<long long> timestamps = GetLastNBlockTimestamps(adjustment_interval);
        if (timestamps.size() > 1) {

            long long lastBlockTimestamp = timestamps.back();
            long long firstBlockTimestamp = timestamps.front();

            long long totalTimeTakenSeconds = lastBlockTimestamp - firstBlockTimestamp;
            double totalExpectedTime = 60.0 * adjustment_interval;

            double newDifficulty = difficulty * (totalExpectedTime / totalTimeTakenSeconds);

            if (newDifficulty > difficulty + 1) {
                newDifficulty = difficulty + 1;
            } else if (newDifficulty < difficulty * 0.75) {
                newDifficulty = difficulty * 0.75;
            }

            this->difficulty = newDifficulty;

            std::cout << "=====================[ MINER INFO ]========================" << std::endl;
            std::cout << "Last timestamp: " << lastBlockTimestamp << " First Timestamp:" << firstBlockTimestamp << std::endl;
            std::cout << "Time taken for " << adjustment_interval << " blocks: " << totalTimeTakenSeconds << " seconds" << std::endl;
            std::cout << "Expected total time for interval: " << totalExpectedTime << " seconds" << std::endl;
            std::cout << "Ratio (expected/actual): " << totalExpectedTime / totalTimeTakenSeconds << std::endl;
            std::cout << "===========================================================" << std::endl;
        }
    }
}




std::vector<long long> Blockchain::GetLastNBlockTimestamps(int n) const {
    std::vector<long long> timestamps;
    int start = std::max(0, this->blocks.Size() - n);
    for (int i = start; i < this->blocks.Size(); ++i) {
        timestamps.push_back(blocks.Get(i)->GetTimestamp());
    }
    return timestamps;
}

std::vector<long long> Blockchain::CalculateTimeDifferences(const std::vector<long long>& timestamps) const {
    std::vector<long long> timeDifferences;
    for (size_t i = 1; i < timestamps.size(); ++i) {
        timeDifferences.push_back(timestamps[i] - timestamps[i - 1]);
    }
    return timeDifferences;
}

double Blockchain::GetAverageTime(const std::vector<long long>& values) const {
    if (values.empty()) return 0;
    long long sum = 0;
    for (const auto& value : values) {
        sum += value;
    }
    return static_cast<double>(sum) / values.size();
}


IMempool *Blockchain::GetMempool() const {
    return this->mempool;
}

File: network/blockchain.ts
import { IBlock } from './block.ts'
import { LinkedList } from './linkedlist.ts';
import { IMempool } from './mempool.ts';

export interface IBlockchain {
	chain: string; // ex: main, testnet
	difficulty: number;
	blocks: LinkedList<IBlock>;
	mempool: IMempool;

	CheckChainIntegrity(): boolean;
	AddBlock(block: IBlock): void;
	GetLastChainBlock(): IBlock | null;
}

export class Blockchain implements IBlockchain{
	chain: string;
	blocks: LinkedList<IBlock>;
	difficulty: number;
	mempool: IMempool;

	private adjustment_interval: number;
	

	constructor(chain_name: string, mempool: IMempool){
		this.chain = chain_name;
		this.difficulty = 1;
		this.blocks = new LinkedList<IBlock>();
		this.adjustment_interval = 2016;
		this.mempool = mempool;
	}
	
	AddBlock(block: IBlock){
		if(block.IsValid()){
			this.blocks.Add(block);
			this.AdjustDifficulty();
		}
	}

	CheckChainIntegrity(): boolean{
		let status = true;
			
		for(let i = 0; i < this.blocks.Size(); i++){
			const currentBlock = this.blocks.Get(i);
			if(currentBlock == null || (currentBlock?.value != null && !currentBlock?.value.IsValid())){
				status = false;
				break;
			}
		}

		return status;

	}
	
	GetLastChainBlock(): IBlock | null{
		if(this.blocks.Size() === 0){
			return null;
		}
		return this.blocks.Get(this.blocks.Size()-1)?.value || null;
	}

	public DisplayChain(): void {
        	for (let i = 0; i < this.blocks.Size(); i++) {
	        	const block = this.blocks.Get(i)?.value;
        		if (block) {
                		console.log(`Block #${i} => Hash: ${block.block_hash}, Previous Hash: ${block.previous_block_hash}`);
            		}
        	}
    	}


	private AdjustDifficulty() {
        	const blockCount = this.blocks.Size();

        	if (blockCount % this.adjustment_interval === 0) {
	        	const timestamps = this.GetLastNBlockTimestamps(this.adjustment_interval);
            		if (timestamps.length > 1) {
                		const timeDifferences = this.CalculateTimeDifferences(timestamps);
                		const medianActualTime = this.GetMedian(timeDifferences);
                		const expectedTime = 60; // 1 block per minute in seconds (target time per block)

                		const newDifficulty = this.difficulty * (expectedTime / medianActualTime);
				
				// prevent the increase of more than 100%
				if(newDifficulty - this.difficulty > 1){
					this.difficulty++;
				}
                		console.log(`New Difficulty: ${this.difficulty}`);
            		}
        	}	
    	}
	private GetLastNBlockTimestamps(n: number): number[] {
        	const timestamps: number[] = [];
        	const start = Math.max(0, this.blocks.Size() - n);
        	for (let i = start; i < this.blocks.Size(); i++) {
            		const block = this.blocks.Get(i)?.value;
            		if (block) {
                		timestamps.push(block.timestamp);
            		}
        	}
        	return timestamps;
	}

    	private CalculateTimeDifferences(timestamps: number[]): number[] {
        	const timeDifferences: number[] = [];
        	for (let i = 1; i < timestamps.length; i++) {
            		timeDifferences.push(timestamps[i] - timestamps[i - 1]);
        	}
        	return timeDifferences;
    	}

    	private GetMedian(values: number[]): number {
        	values.sort((a, b) => a - b); // Sort values in ascending order
        	const middle = Math.floor(values.length / 2);

        	if (values.length % 2 === 0) {
            		// If even, return the average of the two middle values
            		return (values[middle - 1] + values[middle]) / 2;
        	} else {
            		// If odd, return the middle value
            		return values[middle];
        	}
    	}

}
